Feature: experimental meson buildsystem (#1528)
blackmagic-debug/blackmagic: README.md

[![Discord](https://img.shields.io/discord/613131135903596547?logo=discord)](https://discord.gg/P7FYThy)

> # Black Magic Probe

Author: Ah, I was not aware of that rule, I moved the badge to the top because I think it looks better there, it looked to me like it was breaking the flow between the project name and the description

We can move it to additional resources? though that would make it *slightly* harder to spot

> # Black Magic Probe

Reviewer: It does, but we see what you mean about flow - think this needs further reworking with build badges and the like, but that's for another change and time - not this round.

It looks fine in the resources section, so going with that for now.

70      {
```

### Black Magic Debug App

> You can also build the Black Magic Debug suite as a PC program called Black Magic Debug App (BMDA)

Reviewer 2: ```suggestion
You can also build the Black Magic Debug suite as a PC program called Black Magic Debug App (BMDA).
```

Black Magic Debug App also provides for Flashing, reading and verification of a binary file,
by default starting at lowest flash address. The `-t` argument displays information about the
connected target. Use `-h`/`--help` to get a list of supported options.

#### OS specific remarks

> On BSD/Macos, using dev/tty.usbmodemXXX should work but unresolved discussions indicate a hanging open() call on the second invocation. If that happens, try with cu.usbmodemXXX.

Reviewer 2: Fixing some typos in the Apple OS naming, the BSD system family (did you intend a specific one?), and the styling of the device and function names:

```suggestion
On *BSD and macOS, using `/dev/tty.usbmodemXXX` should work but unresolved discussions indicate a hanging `open()` call on the second invocation. If that happens, try with `/dev/cu.usbmodemXXX`.
```

> On BSD/Macos, using dev/tty.usbmodemXXX should work but unresolved discussions indicate a hanging open() call on the second invocation. If that happens, try with cu.usbmodemXXX.

Author: This block I just moved verbatim, I don't know about MacOS, so I did not dive into it. Those do look like reasonable corrections/improvements, and this is a good time to apply them.

## Getting started

### Requirements

When developing this project, the following tools are required:

 * [git](https://git-scm.com)
 * [meson](https://mesonbuild.com/Getting-meson.html) (version 0.61 or greater)
 * [ninja](https://ninja-build.org)

Addtionally, depending on if you want to build/work on the firwmare or Black Magic Debug App (BMDA), you also need:

#### Black Magic Debug Firmware

 * [`arm-none-eabi-gcc`](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/downloads) (from ARM, self-built and distro-supplied is currently broken)

If you wish to use the older [gnu-rm](https://developer.arm.com/downloads/-/gnu-rm) ARM toolchain, this is fine and works well.

> If you have a toolchain from other sources and find problems, check if it is an issue with your toolchain, if not open an issue or better provide a pull request with a fix.

Author: Thanks

#### Black Magic Debug App

 * [GCC](https://gcc.gnu.org) or Clang (Clang is not strictly officially supported)

If you wish to enable support for 3rd party probes, and not only the official native hardware, you will also need:

 * libusb

>  * libftdi

Author: I just copied them verbatim, applied.

>  * libftdi

Reviewer: Yeah, that's entirely fair - figured as we're moving things around and all, we should use the opportunity to fix some cruft 🙂

>  * evt

Author: It did look unfamiliar, but I just copied them verbatim, removed.

 * libhidapi

### Building

> The project is configured and built using the meson buildsystem, you will need to create a build directory and then configure the build depending on what you want.

Reviewer 2: ```suggestion
The project is configured and built using the Meson build system, you will need to create a build directory and then configure the build depending on what you want.
```

The project has a couple of build options you may configure, which affect the resulting binaries, this is how you will configure the firmware for you respective probe, or enable certain features.

A non-exhaustive list of project options:
 * probe: Hardware platform where the BMD firmware will run
 * targets: Enabled debug targets
 * debug_output: Enable debug output (for debugging the BMD stack, not debug targets)
 * rtt_support: Enable RTT (Real Time Transfer) support

You may see all available project options and valid values under  `Project options` in the output of the `meson configure` command.

> The first configuration commands are ran from the root of your clone of the blackmagic repository

Reviewer 2: ```suggestion
The first configuration commands are ran from the root of your clone of the `blackmagic` repository:
```

``` sh
cd /path/to/repository/blackmagic
```

All following commands are ran from inside the build directory.

#### Building Black Magic Debug Firmware

> To build the firmware you need to configure the probe hardware the firmware will run on, as well as a cross-compilation toolchain used.

Reviewer 2: ```suggestion
To build the firmware you need to configure the probe hardware the firmware will run on, as well as the cross-compilation toolchain to be used.
```

For convenience, a cross-file for each supported hardware probe is available which provides a sane default configuration for it.

The build configuration command for the native probe may look like:
``` sh
meson setup build --cross-file cross-file/native.ini

> ```

Author: I did want that be implied somehow, I was hoping the `Changing the build configuration` served the purpose, so I did not have to add too much possibly overwhelming information. But I suppose a note about it here might be helpful

Alternatively (for advanced users), if you wish to configure manually, for instance while writing support for a new probe, or a different toolchain, you can run:
``` sh
meson setup build --cross-file cross-file/arm-none-eabi.ini -Dprobe=native
```

You now should have a `build` directory from where you can build the firmware, this is also where your binary files will appear.

The command `meson compile` will build the default targets:
```sh
cd build
meson compile

> ```

Reviewer 2: No, as any possibly advanced (or clueless) user can say `--backend=vs2010` or `--backend=xcode` and this will not work. An alternative would be the oneliner `meson compile -C build`.

> ```

Reviewer: Oh, true - it's only equivalent on a typical platform and with defaults for the cross-build.. hmm 🤔. We'll let Perigoso decide what they want to do here then and if they should ignore us or write some sort of note

> ```

Author: I chose this because we know it will always just work, even if slightly more verbose, any user comfortable with meson will know they can call ninja directly. It's also clear what the command does for any user, and means we are always interacting with same tool, instead of introducing another one which could possibly induce confusion.

> ```

Author: I chose this because we know it will always just work, even if slightly more verbose, any user comfortable with meson will know they can call ninja directly. It's also clear what the command does for any user, and means we are always interacting with same tool, instead of introducing another one which could possibly induce confusion.

> ```

Reviewer: Yep, ok, we'll accept that rationale. Fair enough!

> Ypu should now see the resulting binaries in `build`, in this case:

Author: Thanks

> ```
> blackmagic_native_firmware.bin
> blackmagic_native_firmware.elf
> ```

Reviewer 2: ```suggestion
- `blackmagic_native_firmware.bin`
- `blackmagic_native_firmware.elf`
```

These are the binary files you will use to flash to your probe.

##### region `rom' overflowed

> It may happen, while working with non default configurations or the project latest version from git, that the firmware does not fit in the available space for the configured probe, this could look something like:

Reviewer 2: ```suggestion
It may happen, while working with non default configurations or the project's latest version from Git, that the firmware does not fit in the available space for the configured probe, this could look something like:
```

```
arm-none-eabi/bin/ld: region `rom' overflowed by 4088 bytes
Memory region         Used Size  Region Size  %age Used
             rom:      135160 B       128 KB    103.12%
             ram:        5708 B        20 KB     27.87%
collect2: error: ld returned 1 exit status
```

> This is not unexpected, as some hardware probe have limited space for firmware, you can get around it by disabling some features or target support:

Reviewer 2: ```suggestion
This is not unexpected, as some hardware probe have limited space for firmware. You can get around it by disabling some features or targets support:
```

``` sh

> meson setup --reconfigure -Dtargets=cortexm,stm -Drtt_support=false

Reviewer: `meson configure -Dtargets=[...]` is preferred  (by us) honestly - and iwrc this is internally rewritten to this form. Trying to describe to users why they're having to do this wordier way is going to be troublesome.

Author: That would be the more correct way to do this, and is my preference too; But while testing the instructions to make sure everything was correct, I noticed that for some reason it did not seem to apply the change properly, the summary would still display the old value...
I did not have time to dive into it, if it's a bug with meson in general, or with the summary feature (or a misshap on my side), but I opted for the command that does work instead, at least for now until we can verify the behavior (If you could check on your system it would be welcome).

Reviewer: We have tested on our system and it works perfectly for us - though we are on Meson 1.0.1 right now, so it's possible there's a bug in a newer Meson. Think we shouldn't be trying to work around that though and should write things the correct way here and report a bug instead. Let's fix the tool in preference 🙂

Author: Ah, great, I was afraid it might be something that affected older versions too, I'm on v1.3.1, I'll investigate this further, but revert these to the more correct command in the meantime

```

#### Building Black Magic Debug App

> The Black Magic Debug App (BMDA) is always build by default, even for firmware builds, so long as all it's dependencies are found, you can find the executable under the build directory, named simply `blackmagic`

Reviewer 2: ```suggestion
The Black Magic Debug App (BMDA) is always built by default, even for firmware builds. So long as all its dependencies are found, you can find the executable under the `build` directory, named simply `blackmagic`.
```

> If you wish to build only the BMDA, you can set the hardware `probe` option to an empty string `-Dprobe=''`, this is the default value:

Reviewer 2: ```suggestion
If you wish to build only BMDA, you can set the hardware `probe` option to an empty string `-Dprobe=''`, this is the default value:
```

``` sh
meson setup build
```

You now should have a `build` directory from where you can build the app, this is also where your executable will appear.

The command `meson compile` will build the default targets:
```sh
cd build
meson compile
```

> Ypu should now see the resulting executable in `build`:

Reviewer 2: ```suggestion
You should now see the resulting executable in `build`:
```

```
blackmagic
```

> Most hosted building is done on and for Linux. BMP-hosted for windows can also be build with Mingw on Linux.

Reviewer: This line can be dropped as it's wrong and has been for ages. BMDA on Windows (the proper name for it) should be done with MSYS2, and we do the full spread of builds for all 3 big name OSes so they're all first-party supported now.

If MinGW were to be named it's purely as a curious extra method. We'd ideally prefer users to use `xwin` to get the headers and use Clang-cl to do a UCRT64 cross build if going down this route. We can provide a cross-file in a follow-up PR that provides the proper environment for doing this as we wrote a Meson configuration for this for bmpflash cross-builds.

Author: Ah, I just copied it over without paying any attention to it, I'm unfamiliar with the windows workflow anyway. removed.

### Changing the build configuration

You may change the configuration at any time after configuring the project, if may also change the default options while configuring for the first time.

Changing options at configure time:
``` sh
meson setup build -Ddebug_output=true
```

> Changin options after configuration:

Reviewer: `Changing`

Same comment for prefering `meson configure` to `meson setup --reconfigure`.

Author: Thanks, fixed the typo

``` sh
meson setup --reconfigure -Dtargets=cortexm,stm
```

> Keep in mind that when changing the configured probe, the other default options will not change, and you may end up with a configuration that does not make sense for it, so in such cases it's best to use the *cross-file* for the probe, not just change the `probe` option.

Reviewer 2: ```suggestion
Keep in mind that when changing the configured probe, the other default options will not change, and you may end up with a configuration that does not make sense for it. In such cases it's best to use the *cross-file* for the probe, not just change the `probe` option.
```

``` sh

> meson setup --reconfigure --cross-file=../cross-file/bluepill.ini

Reviewer: This is fine though as that's the only way we know of to change the cross-file - unless `meson configure` allows it now.

Author: AFAIK, to change the cross-file this is the only supported command

```

You can have multiple build directories! So if you are regularly building firmware for multiple probes we would recomend keeping an individual build directory configured for each one.

### Working with an existing clone (used before the new meson buildsystem was introduced)

If you are working with an existing clone of the project where you used the old `make` buildsystem, or simply initialized the submodules, you may encounter issues while trying the new `meson` workflow, this may look something like:
```
fatal: Fetched in submodule path 'deps/libopencm3', but it did not contain bed4a785eecb6c9e77e7491e196565feb96c617b. Direct fetching of that commit failed.
```

To get around this try running the following commands:
``` sh
git submodule deinit --force --all
meson subprojects purge --confirm

> ```

Reviewer: "
Or for a less destructive method (if you've got customisations and patches in your copy),

```sh
cd deps/libopencm3
git remote set-url origin https://github.com/blackmagic-debug/libopencm3
cd ../..
git submodule update
```
"?

Author: I forgot to warn about lost changes (this is effectively the same as the warning I used to show in the buildsystem with libopecm3), it is a bit nuclear, perhaps not that great an idea to recommend, I'll try reworking it a little

Reviewer: Yep, looks much better 👍🏼

> If this does not resolve the issue, consider making a fresh clone of the repository.

Reviewer: Doing `git remote set-url` in the submodule is guaranteed to work, we'd prefer not to encourage this mentality.

 * [ninja](https://ninja-build.org)

Additionally, depending on if you want to build/work on the firmware or Black Magic Debug App (BMDA), you also need:

#### Black Magic Debug Firmware

 * [`arm-none-eabi-gcc`](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/downloads) (from ARM, self-built and distro-supplied is currently broken)

If you wish to use the older [gnu-rm](https://developer.arm.com/downloads/-/gnu-rm) ARM toolchain, this is fine and works well.

If you have a toolchain from other sources and find problems, check if it's an issue with your toolchain first, and if not then open an issue, or better, submit a pull request with a fix.

#### Black Magic Debug App

 * [GCC](https://gcc.gnu.org) or Clang (Clang is not strictly officially supported)

If you wish to enable support for 3rd party probes, and not only the official native hardware, you will also need:

 * libusb1
 * libftdi1
 * libhidapi

### Building

The project is configured and built using the Meson build system, you will need to create a build directory and then configure the build depending on what you want.

> The project has a couple of build options you may configure, which affect the resulting binaries, this is how you will configure the firmware for you respective probe, or enable certain features.

Reviewer 2: ```suggestion
The project has a couple of build options you may configure which affect the resulting binaries. This is how you will configure the firmware for your respective probe, or enable certain features.
```

A non-exhaustive list of project options:

>  * probe: Hardware platform where the BMD firmware will run
>  * targets: Enabled debug targets
>  * debug_output: Enable debug output (for debugging the BMD stack, not debug targets)
>  * rtt_support: Enable RTT (Real Time Transfer) support
> 

Reviewer 2: ```suggestion
 * `probe`: Hardware platform where the BMD firmware will run
 * `targets`: Enabled debug targets
 * `debug_output`: Enable debug output (for debugging the BMD stack, not debug targets)
 * `rtt_support`: Enable RTT (Real Time Transfer) support

```

Changing options at configure time:
```sh
meson setup build -Ddebug_output=true
```

Changing options after configuration:
```sh
meson configure build -Dtargets=cortexm,stm
```
Alternatively, you can use the equivalent but more verbose `meson setup` command:
```sh
meson setup build --reconfigure -Dtargets=cortexm,stm
```

Keep in mind that when changing the configured probe, the other default options will not change, and you may end up with a configuration that does not make sense for it. In such cases it's best to use the *cross-file* for the probe, not just change the `probe` option. For this you will need to use the `meson setup` command again:
```sh
meson setup build --reconfigure --cross-file=cross-file/bluepill.ini
```

When changing options after configuration, you may omit the argument `build` if you are running the command from within the `build` directory.

You can have multiple build directories! So if you are regularly building firmware for multiple probes we would recommend keeping an individual build directory configured for each one.

### Working with an existing clone (used before the new meson build system was introduced)

> If you are working with an existing clone of the project where you used the old `make` build system, or simply initialized the submodules, you may encounter issues while trying the new `meson` workflow, this may look something like:

Reviewer 2: ```suggestion
If you are working with an existing clone of the project where you used the old Makefile build system, or simply initialized the submodules, you may encounter issues while trying the new Meson workflow, this may look something like:
```

```
fatal: Fetched in submodule path 'deps/libopencm3', but it did not contain bed4a785eecb6c9e77e7491e196565feb96c617b. Direct fetching of that commit failed.
```

To get around this try running the following commands:
```sh
cd deps/libopencm3
git remote set-url origin https://github.com/blackmagic-debug/libopencm3
cd ../..
git submodule update
```

> Alternatively, the nuclear option:
> *BEWARE, THIS WILL ERASE ANY CHANGES THAT HAVE NOT BEEN PUSHED UPSTREAM*

Reviewer 2: ```suggestion
Alternatively, the nuclear option (**BEWARE, THIS WILL ERASE ANY CHANGES THAT HAVE NOT BEEN PUSHED UPSTREAM**):
```

Black Magic Debug App also provides for Flashing, reading and verification of a binary file,
by default starting at lowest flash address. The `-t` argument displays information about the
connected target. Use `-h`/`--help` to get a list of supported options.

#### OS specific remarks

On *BSD and macOS, using `/dev/tty.usbmodemXXX` should work but unresolved discussions indicate a

> hanging `open()` call on the second invocation. If that happens, try with `/dev/cu.usbmodemXXX`.

Reviewer 3: The default recommendation for at least macOS should be the `/dev/cu.usbmodemXXX`. The reason are the caches which are present in the tty version that cause hangups and all kinds of trouble. People should by default use `cu` devices.

Reviewer: That should be taken care of now - please give the new wording a check and resolve this if you're happy.

Reviewer 3: Great! Thanks :)

## Getting started

### Requirements

When developing this project, the following tools are required:

* [git](https://git-scm.com)
* [meson](https://mesonbuild.com/Getting-meson.html) (version 0.61 or greater)
* [ninja](https://ninja-build.org)

Additionally, depending on if you want to build/work on the firmware or Black Magic Debug App (BMDA), you also need:

#### Black Magic Debug Firmware

* [`arm-none-eabi-gcc`](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/downloads) (from ARM, self-built and distro-supplied is currently broken)

If you wish to use the older [gnu-rm](https://developer.arm.com/downloads/-/gnu-rm) ARM toolchain,
this is fine and works well.

If you have a toolchain from other sources and find problems, check if it's an issue with your toolchain first,
and if not then open an issue, or better, submit a pull request with a fix.

#### Black Magic Debug App

> * [GCC](https://gcc.gnu.org) or Clang (Clang is not strictly officially supported)

Reviewer 3: I think it should just say "not officially supported", until that changes.
